stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,449 @@
+// For the code in hcubature.hpp, the package JuliaMath/HCubature.jl
+// written in Julia by Steven G. Johnson served as a template.
+// It comes with the following MIT "Expat" license:
+//
+// Copyright (c) 2017: Steven G. Johnson.
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files
+// (the "Software"), to deal in the Software without restriction,
+// including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software,
+// and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+
+#ifndef STAN_MATH_PRIM_FUNCTOR_HCUBATURE_HPP
+#define STAN_MATH_PRIM_FUNCTOR_HCUBATURE_HPP
+
+#include <stan/math/prim/fun/choose.hpp>
+#include <stan/math/prim/fun/isfinite.hpp>
+#include <stan/math/prim/fun/fabs.hpp>
+#include <stan/math/prim/fun/fmax.hpp>
+#include <stan/math/prim/fun/max.hpp>
+#include <queue>
+#include <tuple>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+static constexpr double xd7[8] = {-9.9145537112081263920685469752598e-01,
+                                  -9.4910791234275852452618968404809e-01,
+                                  -8.6486442335976907278971278864098e-01,
+                                  -7.415311855993944398638647732811e-01,
+                                  -5.8608723546769113029414483825842e-01,
+                                  -4.0584515137739716690660641207707e-01,
+                                  -2.0778495500789846760068940377309e-01,
+                                  0.0};
+
+static constexpr double wd7[8] = {2.2935322010529224963732008059913e-02,
+                                  6.3092092629978553290700663189093e-02,
+                                  1.0479001032225018383987632254189e-01,
+                                  1.4065325971552591874518959051021e-01,
+                                  1.6900472663926790282658342659795e-01,
+                                  1.9035057806478540991325640242055e-01,
+                                  2.0443294007529889241416199923466e-01,
+                                  2.0948214108472782801299917489173e-01};
+
+static constexpr double gwd7[4] = {1.2948496616886969327061143267787e-01,
+                                   2.797053914892766679014677714229e-01,
+                                   3.8183005050511894495036977548818e-01,
+                                   4.1795918367346938775510204081658e-01};
+
+inline void combination(std::vector<int>& c, const int& dim, const int& p,
```

Comment:
Can you add documentation for each of these internal functions? At the moment it might be a bit hard for future readers/users to parse what each is supposed to do

Reply:
Thanks @andrjohns! I added a documentation.